include/pybind11/functional.h

```diff
@@ -38,19 +38,30 @@ struct type_caster<std::function<Return(Args...)>> {
 
         auto func = reinterpret_borrow<function>(src);
 
-        /*
-           When passing a C++ function as an argument to another C++
-           function via Python, every function call would normally involve
-           a full C++ -> Python -> C++ roundtrip, which can be prohibitive.
-           Here, we try to at least detect the case where the function is
-           stateless (i.e. function pointer or lambda function without
-           captured variables), in which case the roundtrip can be avoided.
-         */
+/*
+   When passing a C++ function as an argument to another C++
+   function via Python, every function call would normally involve
+   a full C++ -> Python -> C++ roundtrip, which can be prohibitive.
+   Here, we try to at least detect the case where the function is
+   stateless (i.e. function pointer or lambda function without
+   captured variables), in which case the roundtrip can be avoided.
+ */
+#ifndef BAD
         if (auto cfunc = func.cpp_function()) {
             auto *cfunc_self = PyCFunction_GET_SELF(cfunc.ptr());
-            if (isinstance<capsule>(cfunc_self)) {
+            if (cfunc_self == nullptr) {
+                PyErr_Clear();
+            } else if (isinstance<capsule>(cfunc_self)) {
                 auto c = reinterpret_borrow<capsule>(cfunc_self);
-                auto *rec = (function_record *) c;
+
+                function_record *rec = nullptr;
+                // THIS IS THE PROBLEM: UNSAFE NO WAY TO FIGURE OUT IF CAPSULE IS FOREIGN
+                // We would need to refactor to store a special string such as
+                // pybind11_function_record Doing so is almost certainly an ABI break though Best
+                // we can do without an ABI break is ignore named capsules
+                if (c.name() == nullptr) {
+                    rec = static_cast<function_record *>(c);
```

Comment:
@rwgk Why does a static_cast even work here and not give a compiler error. Do we define a custom conversion somewhere?

Reply:
Looking (Python 3.9) to see if we can find an explanation:

```
#define PyCFunction_GET_SELF(func) \
        (((PyCFunctionObject *)func) -> m_ml -> ml_flags & METH_STATIC ? \
         NULL : ((PyCFunctionObject *)func) -> m_self)
```

Looks like the type of `cfunc_self` is `PyCFunctionObject *`.

Looking that up:
```
typedef struct {
    PyObject_HEAD
    PyMethodDef *m_ml; /* Description of the C function to call */
    PyObject    *m_self; /* Passed as 'self' arg to the C func, can be NULL */
    PyObject    *m_module; /* The __module__ attribute, can be anything */
    PyObject    *m_weakreflist; /* List of weak references */
    vectorcallfunc vectorcall;
} PyCFunctionObject;
```

We're casting that to to a `capsule`:

```
auto c = reinterpret_borrow<capsule>(cfunc_self);
```
With

https://github.com/pybind/pybind11/blob/fab1eebe2c4c52e0abac249f3d058787bc83b5ec/include/pybind11/pytypes.h#L1803

Then here the `capsule` to a `function_record *`:
```
static_cast<function_record *>(c);
```

With

https://github.com/pybind/pybind11/blob/fab1eebe2c4c52e0abac249f3d058787bc83b5ec/include/pybind11/attr.h#L188

And that works? — I don't know how. My way of finding out would be to reduce the code.

Reply:
> @rwgk Nah, we are just extracting m_self from the function. We set that to be a py capsule in other parts of the code. I don't know why the static_cast works from a py::capsule works though.

Oh, yes, sorry I skipped that. So it makes sense that we cast from `PyObject *` via `handle` to `capsule`, but how from `capsule` to `function_record *` I also still don't know.

Reply:
> I don't know why the static_cast works from a py::capsule works though.

`py::capsule` has an implicit `operator T*() const` that looks like it will happily convert to any pointer type you like (and even does so implicitly).  That's... a bit ick.

Reply:
@rwgk Well, we could implicitly convert without a static cast, but that would be even more surprising. I suppose we could also add a special explicit cast method to capsules and deprecate the implicit one somehow? That sounds like another PR though.

Reply:
Based on glancing through this code

* https://github.com/pybind/pybind11/blob/17c1e27b3d202aaab6d5443b1800124bbf24d9ca/include/pybind11/pytypes.h#L1866-L1880

I'm thinking `rec = c.get_pointer<function_record>();` may work.